### PR TITLE
Implement leader election algorithm for the snapshots

### PR DESCRIPTION
**Pull request by vtleonardo**
## Scope

This PR is:
- enabling the leader election of validators to perform the snapshot in the golang and in the smart contracts. The leader election should mitigate race conditions between the validators during the snapshot intervals. The leader chosen to do the snapshot will be rewarded once the snapshot is committed to ethereum (this will be implemented in the near future). 
- cleaning up the leader election logic on golang to work for the ETHDKG and snapshot tasks. 
- Fixing unit tests that started to fail due to the changes

TODO:
- [ ] Compute the snapshot desperation delay in a better way (integrating with the dynamic system)
- [ ] Reward validators that perform the snapshot

### CODE REVIEW

File: layer1/executor/tasks/dkg/mpk_submission.go [hunk 1]
```diff
@@ -125,7 +126,22 @@ func (t *MPKSubmissionTask) Execute(ctx context.Context) (*types.Transaction, *t
...
 	client := t.GetClient()
-	if !utils.AmILeading(client, ctx, logger, int(t.GetStart()), t.StartBlockHash.Bytes(), dkgState.NumberOfValidators, dkgState.Index) {
+	isLeading, err := utils.AmILeading(
+		client,
+		ctx,
+		logger,
+		int(t.GetStart()),
+		t.StartBlockHash.Bytes(),
+		dkgState.NumberOfValidators,
+		dkgState.Index-1,
```
Comment (diligentis) [inline layer1/executor/tasks/dkg/mpk_submission.go:136]: Why the -1 here?

File: layer1/executor/tasks/dkg/mpk_submission.go [hunk 2]
```diff
@@ -125,7 +126,22 @@ func (t *MPKSubmissionTask) Execute(ctx context.Context) (*types.Transaction, *t
...
+		constants.ETHDKGDesperationFactor,
+		constants.ETHDKGDesperationDelay,
+	)
+	if err != nil {
+		return nil, tasks.NewTaskErr("error getting eth height for leader election", true)
```
Comment (diligentis) [inline layer1/executor/tasks/dkg/mpk_submission.go:141]: `err` gets swallowed. Maybe log it before returning a different error?

File: layer1/executor/tasks/dkg/tests/completion_test.go [hunk 3]
```diff
@@ -6,14 +6,15 @@ import (
...
+	"github.com/alicenet/alicenet/constants"
 	"github.com/alicenet/alicenet/layer1/executor/tasks/dkg"
 	"github.com/alicenet/alicenet/layer1/executor/tasks/dkg/state"
-	"github.com/alicenet/alicenet/layer1/executor/tasks/dkg/utils"
 	"github.com/alicenet/alicenet/layer1/monitor/objects"
 	"github.com/alicenet/alicenet/layer1/tests"
 	"github.com/alicenet/alicenet/layer1/transaction"
 	"github.com/alicenet/alicenet/logging"
 	"github.com/alicenet/alicenet/test/mocks"
+	gUtils "github.com/alicenet/alicenet/utils"
```
Comment (diligentis) [inline layer1/executor/tasks/dkg/tests/completion_test.go:17]: No need for alias, looks like only util package after the other one got removed.

File: layer1/executor/tasks/dkg/tests/mpk_submission_test.go [hunk 4]
```diff
@@ -43,7 +44,18 @@ func TestMPKSubmission_Group_1_GoodAllValid(t *testing.T) {
...
 		if shouldExecute {
 			txn, taskError := mpkSubmissionTask.Execute(ctx)
-			amILeading := utils.AmILeading(eth, ctx, logger, int(mpkSubmissionTask.GetStart()), mpkSubmissionTask.StartBlockHash[:], n, dkgState.Index)
+			amILeading, err := gUtils.AmILeading(
+				eth,
+				ctx,
+				logger,
+				int(mpkSubmissionTask.GetStart()),
+				mpkSubmissionTask.StartBlockHash[:],
+				n,
```
Comment (diligentis) [inline layer1/executor/tasks/dkg/tests/mpk_submission_test.go:53]: Might make sense to rename `n` to something more descriptive while we're here.

**Review by diligentis — APPROVED** [review]
Just a few comments, otherwise LGTM!

File: layer1/executor/tasks/dkg/mpk_submission.go (hunk 1 above)
Comment (vtleonardo) [reply layer1/executor/tasks/dkg/mpk_submission.go:136]: The ethdkg process is indexed starting at 1, while the vectors that we store on golang are indexed starting at 0. So, I decided to use the index at 0 in the leader election algorithm. I will be adding some comments about this in the code.

**Pull request merged**